### ISSUE TITLE
CD-ing into a newly created dir or recursive MKD on ftp server fail

### ISSUE THREAD

**Issue by calind** [ncw/rclone#2993]
#### What is the problem you are having with rclone?
I'm trying to proxy a Google Cloud Storage bucket over FTP but creating nested directories fail.

#### What is your rclone version (output from `rclone version`)
```
rclone v1.46
- os/arch: darwin/amd64
- go version: go1.11.5
```

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)
MacOS Mojve 10.14.3

####  Which cloud storage system are you using? (eg Google Drive)
Google Cloud Storage

#### The command you were trying to run
```console
$ rclone serve ftp --cache-dir .rclone-cache gcs:XXXXXXX
```

#### Steps to reproduce
Run the above command and then, in another terminal:
```console
$ lftp -f /dev/stdin <<EOF
OPEN localhost:2121
MKD one
MKD one/two
EOF
mkdir: Access failed: 550 Action not taken: file does not exist (one/two)
$ # the second run succeeds
$ lftp -f /dev/stdin <<EOF
OPEN localhost:2121
MKD one
MKD one/two
EOF
```

As you can see, the second run succeeds. Also, I think this error is related to the issue described here:
```console
$ lftp -f /dev/stdin <<EOF
OPEN localhost:2121
MKD one
CD one
EOF
cd: Access failed: 550 Directory change to /one failed: file does not exist
```

#### A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)
```
2019/02/21 19:11:22 DEBUG : rclone: Version "v1.46" starting with parameters ["rclone" "-vv" "serve" "ftp" "--cache-dir" ".rclone-cache" "gcs:calins-wordpress-runtime-playground"]
2019/02/21 19:11:22 DEBUG : Using config file from "/Users/calin/.config/rclone/rclone.conf"
2019/02/21 19:11:22 INFO  : Storage bucket XXXXXXX: poll-interval is not supported by this remote
2019/02/21 19:11:22 DEBUG : Adding path "vfs/forget" to remote control registry
2019/02/21 19:11:22 DEBUG : Adding path "vfs/refresh" to remote control registry
2019/02/21 19:11:22 DEBUG : Adding path "vfs/poll-interval" to remote control registry
2019/02/21 19:11:22 NOTICE: Storage bucket XXXXXXX: Serving FTP on localhost:2121
2019/02/21 19:11:22 INFO  : : Rclone FTP Server listening on 2121
2019/02/21 19:11:24 DEBUG : : NewDriver: Init driver
2019/02/21 19:11:24 DEBUG : : >NewDriver:
2019/02/21 19:11:24 DEBUG : : Init: Init session
2019/02/21 19:11:24 DEBUG : : >Init:
2019/02/21 19:11:24 INFO  : 8b046f6644e855c84637: Connection Established
2019/02/21 19:11:24 INFO  : 8b046f6644e855c84637: < 220 Welcome on Rclone FTP Server
2019/02/21 19:11:24 INFO  : 8b046f6644e855c84637: > FEAT
2019/02/21 19:11:24 INFO  : 8b046f6644e855c84637: < 211 Extensions supported:
 UTF8
 EPSV
 EPRT
2019/02/21 19:11:24 INFO  : 8b046f6644e855c84637: > OPTS UTF8 ON
2019/02/21 19:11:24 INFO  : 8b046f6644e855c84637: < 200 UTF8 mode enabled
2019/02/21 19:11:24 INFO  : 8b046f6644e855c84637: > USER anonymous
2019/02/21 19:11:24 INFO  : 8b046f6644e855c84637: < 331 User name ok, password required
2019/02/21 19:11:24 INFO  : 8b046f6644e855c84637: > PASS ****
2019/02/21 19:11:24 INFO  : 8b046f6644e855c84637: < 230 Password ok, continue
2019/02/21 19:11:24 INFO  : 8b046f6644e855c84637: > PWD
2019/02/21 19:11:24 INFO  : 8b046f6644e855c84637: < 257 "/" is the current directory
2019/02/21 19:11:24 INFO  : 8b046f6644e855c84637: > MKD one
2019/02/21 19:11:24 DEBUG : /one: MakeDir:
2019/02/21 19:11:25 DEBUG : /one: >MakeDir: err = <nil>
2019/02/21 19:11:25 INFO  : 8b046f6644e855c84637: < 257 Directory created
2019/02/21 19:11:25 INFO  : 8b046f6644e855c84637: > MKD one/two
2019/02/21 19:11:25 DEBUG : /one/two: MakeDir:
2019/02/21 19:11:25 DEBUG : /one/two: >MakeDir: err = file does not exist
2019/02/21 19:11:25 INFO  : 8b046f6644e855c84637: < 550 Action not taken: file does not exist
2019/02/21 19:11:25 INFO  : 8b046f6644e855c84637: > QUIT
2019/02/21 19:11:25 INFO  : 8b046f6644e855c84637: < 221 Goodbye
2019/02/21 19:11:25 INFO  : 8b046f6644e855c84637: Connection Terminated
```

**Comment by ncw** [comment]
Ah, this is to do with how directories are handled on gcs/s3/swift/b2 etc.

Since these are essentially key/value stores they don't have a concept of directories.

When you do `mkdir dir1` rclone does nothing on gcs, so the directory will only live until
- something is put into it
- the `--dir-cache-time` expires

So I would expect this to work

```
$ lftp -f /dev/stdin <<EOF
OPEN localhost:2121
MKD one
MKD one/two
EOF
```

However it doesn't...  This does work though

```
$ lftp -f /dev/stdin <<EOF
OPEN localhost:2121
ls
MKD one
MKD one/two
EOF
```

Which indicates that the first MKD didn't bring the directory into the directory cache somehow...

I think there is a bug in the VFS layer...

Try this

https://beta.rclone.org/branch/v1.46.0-045-g17d90351-fix-2993-vfs-mkdir-beta/ (uploaded in 15-30 mins)

It should work, though note empty directories will disappear after the  `--dir-cache-time` expires

**Comment by calind** [comment]
Yes, it fixes the issue (I had to compile the branch from source, at commit 17d90351 because I'm on darwin).

Empty directories dissapearing after `--dir-cache-time` is an expected behaior in this case I think.

**Comment by calind** [comment]
Yes, it fixes the issue (I had to compile the branch from source, at commit 17d90351 because I'm on darwin).

Empty directories dissapearing after `--dir-cache-time` is an expected behaior in this case I think.

**Comment by ncw** [comment]
> Yes, it fixes the issue (I had to compile the branch from source, at commit [17d9035](https://github.com/ncw/rclone/commit/17d90351916ed512d38a9eb724cc9b80490acaf6) because I'm on darwin).

Ah, sorry, the macOS build has been a bit unreliable recently :-(



> Empty directories dissapearing after `--dir-cache-time` is an expected behaior in this case I think.

I think its the best we can do at the moment.

I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.47

Thanks for testing :-)
